### ISSUE TITLE
[swift2objc] Extend Builtin Declarations (mainly builtin protocols)

### ISSUE THREAD

**Issue by nikeokoronkwo** [dart-lang/native#1827]
This issue is to add more builtin declarations to the ones already existing in the swift2objc tool:
A few of the main categories of declarations to add are here

### Inbuilt Protocols
Swift has a set of inbuilt Core Protocols that can be used by classes, structs, other protocols and generics in type constraints. A few of these are: `Hashable`, `Numeric`, `RandomNumberGenerator`, etc.

When making use of such protocols in classes, the protocol implementations are not automatically added to the symbol graph (although they are present, like other inbuilt types like `String`), and so would need to be added as a builtin declaration.

### More Inbuilt Types
A few more types may need to be added, like `Dict`, `Array`, `Set`, etc. Some of these types also make use of [generics](#1814) 

----------------

Hopefully by the end of this issue a good amount of the builtin declarations will be present.
A full list of the builtin types in the swift standard library can be found here: https://developer.apple.com/documentation/swift/swift-standard-library

**Comment by nikeokoronkwo** [comment]
There are a lot of items under the Swift Standard Library, so we might not import all of them. However, it would be helpful to layout the groundwork for such so that if we encounter any libraries containing types not included here, they can easily be added.

This would mean dropping the current `enum` approach for a more inclusive approach that can include other types like protocols, enums and more. By the way, most swift types are actually `struct`s under the hood, and some even have nested declarations (e.g `Swift.UTF16View`), which need to be considered, which is why the `enum` approach would need to be dropped.

**Comment by liamappelbe** [comment]
> This would mean dropping the current `enum` approach for a more inclusive approach that can include other types like protocols, enums and more.

Yeah, I was working on another refactor and found that I needed to change `BuiltInDeclaration` to a class. I'll try and get that refactor into a PR today.

**Comment by liamappelbe** [comment]
> When making use of such protocols in classes, the protocol implementations are not automatically added to the symbol graph (although they are present, like other inbuilt types like `String`), and so would need to be added as a builtin declaration.

What exactly is going on with these protocols? Where are they defined? My hunch is they're not actually built-in, but are defined in Foundation. If that's the case, then instead of adding these protocols as built-ins, we'll be parsing them automatically once https://github.com/dart-lang/native/issues/1774 is implemented.

**Comment by nikeokoronkwo** [comment]
> > When making use of such protocols in classes, the protocol implementations are not automatically added to the symbol graph (although they are present, like other inbuilt types like `String`), and so would need to be added as a builtin declaration.
> 
> What exactly is going on with these protocols? Where are they defined? My hunch is they're not actually built-in, but are defined in Foundation. If that's the case, then instead of adding these protocols as built-ins, we'll be parsing them automatically once #1774 is implemented.

The protocols are actually built in as a part of the [Swift Standard Library](https://developer.apple.com/documentation/swift/swift-standard-library), and are actually separate from `Foundation`.  Some protocols that are in the standard library are: `Collection`, `Hashable`, `Numeric`, `Equatable` and more. 

`Foundation` has its own set of protocols, structs and other declarations. Some protocols in the `Foundation` library are: `DataProtocol`, `NSFastEnumeration`

**Comment by nikeokoronkwo** [comment]
An Idea I thought of for this would be to have the builtin declarations in a separate place in the codebase. 

The important ones like `String`, `Hashable`, `Equatable` and `Bool` can be declared and used as priority, and then fall back into other ones when searching for declarations. The other declarations can be separated into different directories based on topic.

However, this would mean that we would have to make `BuiltInDeclaration` more of a mixin (to denote that this declaration is "built in") and then make Builtin variants of `StructDeclaration` and others, in order to support types like `String.UTF16View`, `Regex<String>` and `Dictionary<String, String>`.

Does the `Any` type appear in symbol graphs?
